src/core/pointcloud/qgspointclouddataprovider.cpp

```diff
@@ -176,3 +183,145 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
+  {
+
+  }
+
+  QVector<QMap<QString, QVariant>> operator()( const IndexedPointCloudNode &n )
+  {
+    QVector<QMap<QString, QVariant>> acceptedPoints;
+    std::unique_ptr<QgsPointCloudBlock> block( mIndex->nodeData( n, mRequest ) );
+
+    if ( !block || pointsCount == mPointsLimit )
+      return acceptedPoints;
+
+    const char *ptr = block->data();
+    QgsPointCloudAttributeCollection blockAttributes = block->attributes();
+    const std::size_t recordSize = blockAttributes.pointRecordSize();
+    mContext.setAttributes( block->attributes() );
+    for ( int i = 0; i < block->pointCount(); ++i )
+    {
+      double x, y, z;
+      _pointXY( mContext, ptr, i, x, y );
+      z = _pointZ( mContext, ptr, i );
+      QgsPointXY pointXY( x, y );
+
+      if ( pointsCount < mPointsLimit && mExtentGeometry.contains( &pointXY ) && mZRange.contains( z ) )
+      {
+        QMap<QString, QVariant> pointAttr = mContext.attributeMap( ptr, i * recordSize, blockAttributes );
+        pointAttr[ QStringLiteral( "X" ) ] = x;
+        pointAttr[ QStringLiteral( "Y" ) ] = y;
+        pointAttr[ QStringLiteral( "Z" ) ] = z;
+        pointsCount++;
+        acceptedPoints.push_back( pointAttr );
+      }
+    }
+    return acceptedPoints;
+  };
+
+  QgsPointCloudRequest &mRequest;
+  QgsPointCloudRenderContext &mContext;
+  QgsGeometry &mExtentGeometry;
+  const QgsDoubleRange &mZRange;
+  QgsPointCloudIndex *mIndex = nullptr;
+  int mPointsLimit;
+  int pointsCount = 0;
+};
+
+QVector<QMap<QString, QVariant>> QgsPointCloudDataProvider::identify(
+                                float maxErrorInMapCoords,
+                                QgsGeometry extentGeometry,
+                                const QgsDoubleRange extentZRange, int pointsLimit )
+{
+  QVector<QMap<QString, QVariant>> acceptedPoints;
+
+  QgsPointCloudIndex *index = this->index();
+  const IndexedPointCloudNode root = index->root();
+
+  QgsRenderContext renderContext;
+
+  QgsRectangle rootNodeExtentMapCoords = index->nodeMapExtent( root );
+// TODO? reproject the root node extent
+//  try
+//  {
+//    rootNodeExtentMapCoords = renderContext.coordinateTransform().transformBoundingBox( index->nodeMapExtent( root ) );
+//  }
+//  catch ( QgsCsException & )
+//  {
+//    QgsDebugMsg( QStringLiteral( "Could not transform node extent to map CRS" ) );
+//  }
+  const float rootErrorInMapCoordinates = rootNodeExtentMapCoords.width() / index->span();
+
+  QVector<IndexedPointCloudNode> nodes = traverseTree( index, root, maxErrorInMapCoords, rootErrorInMapCoordinates, extentGeometry, extentZRange );
+
+  QgsPointCloudAttributeCollection attributeCollection = index->attributes();
+  QgsPointCloudRequest request;
+  request.setAttributes( attributeCollection );
+
+  QgsPointCloudRenderContext context( renderContext, index->scale(), index->offset(), 1.0, 0.0 );
```

Comment:
We still need to remove this from the data provider class -- it's a render specific class which should not be used here